pkg/controllers/management/auth/token_test.go

```diff
@@ -0,0 +1,210 @@
+	"github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3/fakes"
+	"github.com/stretchr/testify/assert"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type tokenTestCase struct {
+	inputToken                  *v3.Token
+	expectedOutputToken         *v3.Token
+	inputUserAttribute          *v3.UserAttribute
+	expectedOutputUserAttribute *v3.UserAttribute
+	enableHashing               bool
+	description                 string
+}
+
+func TestSync(t *testing.T) {
+	tokens := make(map[string]*v3.Token)
+	userAttributes := make(map[string]*v3.UserAttribute)
+
+	testTokenController := TokenController{
+		tokens: &fakes.TokenInterfaceMock{
+			UpdateFunc: func(token *v3.Token) (*v3.Token, error) {
+				tokens[token.Name] = token.DeepCopy()
+				return token, nil
+			},
+			GetFunc: func(name string, opts metav1.GetOptions) (*v3.Token, error) {
+				token, ok := tokens[name]
+				if ok {
+					return token, nil
+				}
+				return nil, errors.NewNotFound(schema.GroupResource{}, name)
+			},
+		},
+		userAttributes: &fakes.UserAttributeInterfaceMock{
+			UpdateFunc: func(userAttribute *v3.UserAttribute) (*v3.UserAttribute, error) {
+				userAttributes[userAttribute.Name] = userAttribute.DeepCopy()
+				return userAttribute, nil
+			},
+		},
+		userAttributesLister: &fakes.UserAttributeListerMock{
+			GetFunc: func(namespace string, name string) (*v3.UserAttribute, error) {
+				userAttribute, ok := userAttributes[name]
+				if ok {
+					return userAttribute, nil
+				}
+				return nil, errors.NewNotFound(schema.GroupResource{}, name)
+			},
+		},
+	}
+	testCases := populateTestCases(tokens, userAttributes)
+	// test legacy finalizer removal
+	for _, testcase := range testCases {
+		if testcase.enableHashing {
+			features.TokenHashing.Set(true)
+		}
+		returnToken, _ := testTokenController.sync(testcase.inputToken.Name, testcase.inputToken)
+		storedToken, _ := testTokenController.tokens.Get(testcase.inputToken.Name, metav1.GetOptions{})
+		assert.Equal(t, returnToken, storedToken, fmt.Sprintf("%s", testcase.inputToken.Name))
+		features.TokenHashing.Set(false)
+		if testcase.enableHashing {
+			assert.NotEqual(t, returnToken.(*v3.Token).Token, testcase.inputToken.Token)
+			assert.True(t, strings.HasPrefix(returnToken.(*v3.Token).Token, "$2"))
+			testcase.expectedOutputToken.Token = ""
+			returnToken.(*v3.Token).Token = ""
+		}
+		assert.Equal(t, testcase.expectedOutputToken, returnToken, fmt.Sprintf("%s", testcase.inputToken.Name))
+		if testcase.inputUserAttribute == nil {
+			continue
+		}
+		returnUserAttribute, _ := testTokenController.userAttributesLister.Get("", testcase.inputUserAttribute.Name)
+		assert.Equal(t, testcase.expectedOutputUserAttribute, returnUserAttribute, fmt.Sprintf("%s", testcase.inputToken.Name))
+	}
+}
+
+func populateTestCases(tokens map[string]*v3.Token, userAttributes map[string]*v3.UserAttribute) []tokenTestCase {
```

Comment:
One test nit: I think that you got all of the standard test cases, but did not include any error test cases (error when updating after removing finalizer, error when updating after hashing, error when refreshing attribute) and the obj == nil case. Since these are of questionable value (I got an error I return an error), you don't have to add them, but I thought I would leave a comment on it just in case.

Reply:
These were all added besides the obj==nil case. Resolving.